Dockerfile

```diff
@@ -52,41 +54,84 @@ RUN ln -s /usr/bin/python3 /usr/bin/python
 
 WORKDIR /edx/app/edx-platform/edx-platform
 
-COPY . /edx/app/edx-platform/edx-platform
-
 ENV PATH /edx/app/edx-platform/nodeenv/bin:${PATH}
 ENV PATH ./node_modules/.bin:${PATH}
 ENV CONFIG_ROOT /edx/etc/
 ENV PATH /edx/app/edx-platform/edx-platform/bin:${PATH}
 ENV SETTINGS production
+RUN mkdir -p /edx/etc/
 
-# TODO: Install requirements before copying in code.
+# Distinguish requirements that come from code within edx-platform
+# from external requirements (specifically, those from pip or GitHub).
+# Then, install just external requirements.
+# This way, the cached installations of external requirements are not
+# busted whenever edx-platform code is changed.
+# TODO this is quite a hack and should be done in `make upgrade` probably.
 RUN pip install setuptools==39.0.1 pip==9.0.3
-RUN pip install -r requirements/edx/base.txt
+COPY requirements/ requirements/
+RUN grep -E    "^-e (common|openedx|lms|cms|\.)(/| )" requirements/edx/base.txt > requirements/edx/base_in_tree.txt
+RUN grep -E -v "^-e (common|openedx|lms|cms|\.)(/| )" requirements/edx/base.txt > requirements/edx/base_not_in_tree.txt
+RUN pip install -r requirements/edx/base_not_in_tree.txt
 
+# Copy just JS requirements and install them.
+COPY package.json package.json
+COPY package-lock.json package-lock.json
+# TODO: Shouldn't we use node==12.11.1?
 RUN nodeenv /edx/app/edx-platform/nodeenv --node=8.9.3 --prebuilt
+RUN npm set progress=false && npm install
 
-RUN npm set progress=false \
-    && npm install
+# Install remaining requirements -- that is, the in-tree ones.
+COPY setup.py setup.py
+COPY common common
+COPY openedx openedx
+COPY lms lms
+COPY cms cms
+RUN pip install -r requirements/edx/base_in_tree.txt
 
-RUN mkdir -p /edx/etc/
+ENV LMS_CFG /edx/etc/lms.yml
+ENV STUDIO_CFG /edx/etc/studio.yml
+COPY lms/devstack.yml /edx/etc/lms.yml
+COPY cms/devstack.yml /edx/etc/studio.yml
```

Comment:
(This is one of the action items in the [post-spike action plan](https://openedx.atlassian.net/wiki/spaces/ENG/pages/1685619678/Post-Spike+Action+Plan?focusedCommentId=1746436117#comment-1746436117) under **Reconcile settings handling between edx-platform and course-discovery PRs.**).

We ought to make settings handling between this and the Discovery PR consistent before merging.